### PR TITLE
Handle missing file while hashing

### PR DESCRIPTION
**Pull request by angelini**
This is to address an issue @airhorns saw in CI.

A file `fsdiff` was trying to hash no longer existed on disk.

The likely scenario is that we marked the file as present while walking the directory and when we later tried to read and hash it, it had been deleted concurrently.

This should be quite rare as we only hash a few exceptional files. I tried to explain this in a comment but there are more details here: https://github.com/gadget-inc/fsdiff/pull/10

I can't think of a reliable way to test this, we would need to somehow modify the filesystem between `Diff`'s read and it's hashing process.

### CODE REVIEW

File: pkg/diff/diff.go [hunk 1]
```diff
@@ -230,16 +230,43 @@ func hashEntry(dir string, entry *pb.Entry) ([]byte, error) {
...
 			}
+
 			entry.Hash = hash
 		}
+
+		summary.Entries[idx] = entry
+		idx++
+	}
+
+	// Remove any dangling elements on the end of the slice
```
Comment (angelini) [inline pkg/diff/diff.go:265]: https://stackoverflow.com/a/20551116

Comment (scott-rc) [reply pkg/diff/diff.go:265]: Damn 🔥

**Review by scott-rc — APPROVED** [review]

**Review by airhorns — APPROVED** [review]
Makes sense to me! I know that one way folks have tested stuff like this in the past is to add a virtualized filesystem that everything operates on, where you don't use the real FS and instead pass around a handle to one. In prod its a real FS, but in tests it's a virtual one that lets you simulate and coordinate and block and do whatever you want with full control over the sequencing. FoundationDB used this technique to mock both the FS and network for their shit and are super into it. In fsdiff's case I think pragmatically its probably not worth it, especially right now, but if we find we keep hitting this kinda stuff we could look into it!

**Comment by angelini** [comment]
> I know that one way folks have tested stuff like this in the past is to add a virtualized filesystem that everything operates on, where you don't use the real FS and instead pass around a handle to one. In prod its a real FS, but in tests it's a virtual one that lets you simulate and coordinate and block and do whatever you want with full control over the sequencing. FoundationDB used this technique to mock both the FS and network for their shit and are super into it. In fsdiff's case I think pragmatically its probably not worth it, especially right now, but if we find we keep hitting this kinda stuff we could look into it!

Ohh that's really cool, going to dig into how FoundationDB did it.

Probably not super worth it right now for `fsdiff` but it might be a good testing ground for it and that would be super useful in DateiLager's test suite.

**Pull request merged**